### DISCUSSION TITLE
Have github action only run on master repo, and not on forks?

### DISCUSSION THREAD

**Discussion by Blakeinstein** [community/community#26409] [Actions]
<p>I have an action that will always fail on forks due to the lack of secrets.</p>
<p>When I set the action permissions to <code>Allow local actions only</code>.</p>
<p>My test action on PRs fails.</p>
<p>How would I about setting up required permissions?<br>
So that, my deploy/update actions run on my own repo only and my test workflow works on PRs?</p>

**Comment by kingthorin** [comment]
<p>If you use the event <code>pull_request_target</code> and actually checkout the PR branch (see below) then your PR workflow should be fine.</p>
<pre><code class="lang-auto">    - name: Checkout
      uses: actions/checkout@v2
      with:
        ref: ${{github.event.pull_request.head.ref}}
        repository: ${{github.event.pull_request.head.repo.full_name}}
</code></pre>
<ul>
<li><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target</a></li>
<li><a href="https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/" rel="noopener nofollow ugc">https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/</a></li>
</ul>
<p>Currently there isn’t a way to control enable/disable state of workflows in forks.</p>

**Comment by Blakeinstein** [comment]
<p>I am checking out the branch… am I supposed to add the refs?</p>
<p>also by pr target… I have</p>
<pre><code class="lang-yaml">on:
  pull_request:
    branches:
      - master
</code></pre>
<p>further more. What setting am I supposed to set?<br>
With local actions only, I get</p>
<pre><code class="lang-bash">actions/checkout@v2 and actions/setup-node@v2-beta are not allowed to be used in Upvision/upvision.github.io. Actions in this workflow must be: within a repository owned by Upvision.
</code></pre>

**Comment by kingthorin** [comment]
<p>Local actions means those that are in your repo. (Of which you probably have none.)</p>
<p>Here’s a PR action that does markdown linting on submissions to help you get a model/understanding of how things fit together: <a href="https://github.com/OWASP/wstg/blob/fd5674975bf20209c13e6a6195744e6cf35d0152/.github/workflows/md-lint-check.yml" rel="noopener nofollow ugc">https://github.com/OWASP/wstg/blob/fd5674975bf20209c13e6a6195744e6cf35d0152/.github/workflows/md-lint-check.yml</a></p>

**Comment by Blakeinstein** [comment]
<p>With said changes. I get<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/2/c/2c1d8b2959f99bd104aabe7250672d04ea0d0ed8.png" data-download-href="/uploads/short-url/6igiWBWAtRzKoMDac8jeOFB4w3e.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/42214598/181097535-2d9ebd19-ad70-4a7c-870c-d3c3d08fc4b6.png" alt="image" data-base62-sha1="6igiWBWAtRzKoMDac8jeOFB4w3e" width="690" height="218" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/c/2c1d8b2959f99bd104aabe7250672d04ea0d0ed8_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">855×271 13.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by kingthorin** [comment]
<p>Okay, so it’s running, that’s great!</p>

**Comment by Blakeinstein** [comment]
<p>Unfortunately no. Its just stuck there.</p>

**Comment by kingthorin** [comment]
<p>Is it a public repo, can you provide a link?</p>

**Comment by Blakeinstein** [comment]
<p>Its <a href="https://github.com/upvision/upvision.github.io" rel="noopener nofollow ugc">https://github.com/upvision/upvision.github.io</a></p>

**Comment by kingthorin** [comment]
<p>The gh-test workflow? (It was last updated 2h ago, but looking at your actions tab hasn’t been triggered in 7h.)</p>
<p>Also your publish action is reporting a syntax issue: <a href="https://github.com/Upvision/upvision.github.io/actions/runs/339580524" rel="noopener nofollow ugc">https://github.com/Upvision/upvision.github.io/actions/runs/339580524</a></p>

**Comment by Blakeinstein** [comment]
<p>Yes I had to run tests locally and force merge, because i was on a schedule. The tests never run and await status to be reported.<br>
I am told this happens when a workflow file is updated, but this never happened to me before when I did update the test workflow, but recently started showing up when I tried to restrict actions in forks, even when actions were not updated.</p>
<p>As for the syntax issue, it has something to do with the env action, as its documentation requires it to be configured that way.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/blakeinstein">@Blakeinstein</a>,</p>
<p>The “<strong>Allow local actions only</strong>” option in the Organization settings means that the <a href="https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/about-actions" rel="noopener nofollow ugc"><strong>actions</strong></a> (don’t confuse with <strong>workflows</strong>) only defined in the repositories within current organization can be used in the workflows within current organization. The workflows within current organization can’t use the actions defined outside of current organization.<br>
That’s why you get the message:</p>
<blockquote>
<p>actions/checkout@v2 and actions/setup-node@v2-beta are not allowed to be used in Upvision/upvision.github.io. Actions in this workflow must be: within a repository owned by Upvision.</p>
</blockquote>
<p>Because the “<strong>checkout</strong>” action and “<strong>setup-node</strong>” action are defined within the “<strong>actions</strong>” organization, not your organization “<strong>Upvision</strong>”.</p>
<p>If you want to use the actions from other organizations, you can:</p>
<ul>
<li>
<p>select “<strong>Allow all actions</strong>”, the actions from anywhere can be used in the workflows within current organization.</p>
</li>
<li>
<p>select “<strong>Allow select actions</strong>”, specify the actions you want to run within current organization. The specified actions can be from anywhere.</p>
</li>
</ul>
<br>
<p>About workflows from fork repository, there also are some option you can use in the Organization settings.<br>
<img src="https://user-images.githubusercontent.com/46307996/181097542-40d47fcd-7223-46b9-8dee-b1fd85468680.png" alt="settings" data-base62-sha1="p9iRqDe6FO4Svhbfo34lzmmbFCg" width="637" height="371"></p>
<br>
<p>To view more details, you can see “<a href="https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-organizations-and-teams/disabling-or-limiting-github-actions-for-your-organization" rel="noopener nofollow ugc"><strong>Disabling or limiting GitHub Actions for your organization</strong></a>”.</p>

**Comment by Blakeinstein** [comment]
<p>I dont want the actions defined on the main repository to run on the forks. I do not want to expose secrets either.</p>

**Comment by Simran-B** [comment]
<p>You’re still confusing actions and workflows. You cannot stop forked repos from causing workflow runs by reducing actions permissions. What you want is to disable <em>Run workflows from fork pull requests</em>, but that isn’t available for regular accounts and it seems to be limited to private repos?</p>
<aside class="quote no-group" data-username="Blakeinstein" data-post="13" data-topic="140840">
<div class="title">
<div class="quote-controls"></div>
 Blakeinstein:</div>
<blockquote>
<p>I dont want the actions defined on the main repository to run on the forks.</p>
</blockquote>
</aside>
<p>They don’t run in forked repos by default, but it can be enabled by the owner of that fork. Note that this is different from pull requests being opened against your repo, which may trigger a workflow run on your end.</p>
<p>What you could do is to only run jobs if the feature branch is in your repo (not in a fork):</p>
<pre><code class="lang-yaml">jobs:
  job_id:
    if: github.event.pull_request.head.repo.full_name == github.repository
    ...
</code></pre>

**Comment by airtower-luna** [comment]
<p>In that case you probably don’t need the <code>pull_request</code> event at all, if you use the <code>push</code> event it’ll apply to all branches in your own repository, too, unless you use a branch filter.</p>
<p>That said, looking at your initial post it sounds like you want to run on PRs, just not the deploy stuff:</p>
<aside class="quote no-group" data-username="Blakeinstein" data-post="1" data-topic="140840">
<div class="title">
<div class="quote-controls"></div>
 Blakeinstein:</div>
<blockquote>
<p>So that, my deploy/update actions run on my own repo only and my test workflow works on PRs?</p>
</blockquote>
</aside>
<p>I think the easiest solution is:</p>
<ol>
<li>Create a test workflow that doesn’t use secrets, and have that trigger on <code>pull_request</code> (and maybe <code>push</code> for you local branches).</li>
<li>Create a separate workflow with <code>push</code> event and branch filter for whatever branch you want to deploy from, and have it do the deploy/update. You’ll probably want to run the tests in that workflow, too, even if it’s a bit of duplication, to make sure only successfully tested code gets deployed.</li>
</ol>

**Comment by Blakeinstein** [comment]
<p>I have a workflow that is triggered on pull requests, and that works as intended. But my main workflow that deploys the webpage, is also being run on forks. Its not an issue as they simply error out because of the lack of secrets. But I’d rather not have the workflows run on forks.</p>

**Comment by Simran-B** [comment]
<p>Adding conditions as I showed should be a viable solution then. Workflows will still be started on forks, but abort early without causing failed workflow runs.</p>

**Comment by s-weigand** [comment, marked as answer]
<p><a class="mention" href="/simran-b">@Simran-B</a> gave the solution you want and need :wink:</p>
<p>To make the condition more portable (copy/paste) in an organization or your own account you can just use the repository owner name as a condition to match.</p>
<p>In your case just replace</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/Upvision/upvision.github.io/blob/41ac1a284e4b3accc09309d852ca3b477d1f395e/.github/workflows/gh-deploy.yml#L10-L11" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/Upvision/upvision.github.io/blob/41ac1a284e4b3accc09309d852ca3b477d1f395e/.github/workflows/gh-deploy.yml#L10-L11" target="_blank" rel="noopener nofollow ugc">Upvision/upvision.github.io/blob/41ac1a284e4b3accc09309d852ca3b477d1f395e/.github/workflows/gh-deploy.yml#L10-L11</a></h4>
<pre class="onebox"><code class="lang-yml"><ol class="start lines" start="10" style="counter-reset: li-counter 9 ;">
<li>jobs:</li>
<li>  publish:</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>with</p>
<pre><code class="lang-yaml">jobs:
  publish:
    if: github.repository_owner == 'Upvision'
</code></pre>
<p>For the forks not to be bothered with the failing workflow they just need to update to your version.</p>

**Comment by Blakeinstein** [comment]
<p>That is exactly what I needed! Thanks!</p>
<p>Sorry for the earlier confusion!</p>

**Comment by devedse** [comment]


**Reply by kingthorin** [reply]
❓

**Reply by devedse** [reply]
Sorry I mistyped a comment in the wrong place

**Comment by jsoref** [comment]
Fwiw, what you're doing isn't really ideal.

Instead of hardcoding "is my owner X", it's friendlier to add a condition that checks for the required secrets. This would allow someone else to provide compatible secrets for testing in their fork.

In the simplest case, it can be something like this:
https://github.com/smallstep/cli/blob/94f3d85d3de96deb01e0dbc1b60596ecda12ac43/.github/workflows/triage.yml#L29-L31

For fancier cases, you can make a step w/ an output, or a job w/ an output and then condition other steps/jobs based on those outputs.

**Reply by thbeu** [reply]
That does not work with the `pull_request` event, but only with the `pull_request_target` event.

**Reply by jsoref** [reply]
Eh? The example is for the issue event, not pull request target.

**Comment by ghogen** [comment]
The way our organization uses forks, we never want forks of our repo to be able to run any GitHub actions. It's not a question of secrets and permissions, it's a more fundamental issue in that forks are only for working on content changes and submitting them (via PRs) to the original repo, they are not for managing anything like tracking issues or running actions. I think a setting is still needed to disallow GH actions on forked repos, not just something in the individual actions. I guess that's a feature request, since I don't think there's any such setting now.

**Reply by jsoref** [reply]
How someone else uses their fork of your repository is not up to you, it's up to them.

If you think that your workflows aren't useful in forks, then your template for workflows should have the code to restrict them to your organization.
